physical/postgresql.go

```diff
@@ -0,0 +1,268 @@
+package physical
+
+import (
+	"database/sql"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"sync"
+	"time"
+)
+
+// PostGresBackend is a physical backend that stores data on postgres database
+type PostGresBackend struct {
+	Url string
+
+	l sync.Mutex
+}
+
+var (
+	Trace   *log.Logger
+	Info    *log.Logger
+	Warning *log.Logger
+	Error   *log.Logger
+)
+
+// newFileBackend constructs a Filebackend using the given directory
+func newPostGresBackend(conf map[string]string) (Backend, error) {
+
+	Trace = log.New(ioutil.Discard, "Trace: ", log.Ldate|log.Ltime|log.Lshortfile)
+	Info = log.New(os.Stdout, "INFO: ", log.Ldate|log.Ltime|log.Lshortfile)
+	Warning = log.New(os.Stdout, "WARNING: ", log.Ldate|log.Ltime|log.Lshortfile)
+	Error = log.New(os.Stdout, "ERROR: ", log.Ldate|log.Ltime|log.Lshortfile)
+
+	Trace.Println("> newPostGresBackend")
+	url, ok := conf["url"]
+	if !ok {
+		return nil, fmt.Errorf("'url' must be set")
+	}
+
+	db, err := sql.Open("postgres", url)
+	if err != nil {
+		Error.Fatal(err)
+	}
+	defer db.Close()
+	sqlStmt := "CREATE EXTENSION if not exists \"uuid-ossp\";"
+	_, err = db.Exec(sqlStmt)
+	if err != nil {
+		Error.Printf("%q: %s\n", err, sqlStmt)
+		Error.Println("< newPostGresBackend(err)")
+		Error.Panic(err)
+	}
+
+	sqlStmt = `
+CREATE TABLE if not exists vault (
+  id UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
+  key TEXT not null,
+  value bytea,
+  created_at TIMESTAMP WITHOUT TIME ZONE NOT NULL,
+  updated_at TIMESTAMP WITHOUT TIME ZONE NOT NULL
+);`
+	_, err = db.Exec(sqlStmt)
+	if err != nil {
+		Error.Printf("%q: %s\n", err, sqlStmt)
+		Error.Println("< newPostGresBackend(err)")
+		Error.Panic(err)
+	}
+
+	Trace.Println("< newPostGresBackend")
+	return &PostGresBackend{Url: url}, nil
+}
+
+func (b *PostGresBackend) Delete(k string) error {
+	Trace.Println("> Delete")
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	db, err := sql.Open("postgres", b.Url)
+	if err != nil {
+		Trace.Println("< Delete error db")
+		Error.Fatal(err)
+	}
+	defer db.Close()
+
+	txn, err := db.Begin()
+	if err != nil {
+		Trace.Println("< Delete error begin")
+		Trace.Fatal(err)
+	}
+
+	stmt, err := txn.Prepare("delete from vault where key = $1")
+	if err != nil {
+		Trace.Println("< Delete error prepare")
+		Error.Fatal(err)
+	}
+
+	_, err = stmt.Exec(k)
+	if err != nil {
+		Trace.Println("< Delete error exec")
+		Error.Fatal(err)
```

Comment:
I'm not sure if the postgres backend should panic and crash the vault instance if it can't run a query! :smile:  Usually go programs tend to return errors the encounter to the caller. Sometimes they [wrap them in another error](https://github.com/hashicorp/vault/blob/e2e077964b5200208a7554963d49f55ec98ade57/physical/consul.go#L103) to give some context as to what area generated the error.